codegen/project_generator.go

```diff
@@ -49,90 +59,102 @@ func NewProjectGenerator(srcPath, projectName string, opts ...ProjectGeneratorOp
-			return fmt.Errorf("creating missing %q folder: %w", g.srcPath, err)
+		if err := os.MkdirAll(projectPath, 0755); err != nil {
+			return fmt.Errorf("creating missing %q folder: %w", projectPath, err)
 		}
 	}
 
-	fmt.Printf("Generating files in %q\n", fullPath)
-
-	srcDir := filepath.Join(fullPath, "src")
-	if _, err := os.Stat(srcDir); errors.Is(err, os.ErrNotExist) {
-		fmt.Printf("Creating missing %q folder\n", srcDir)
-		if err := os.MkdirAll(srcDir, os.ModePerm); err != nil {
-			return fmt.Errorf("creating src directory %v: %w", srcDir, err)
-		}
-	} else {
-		fmt.Println("src directory already exists, skipping")
+	// generate template from ./templates
+	templateFiles, err := template.New("templates").Funcs(utils).ParseFS(templates, "*/*.gotmpl", "*/*/*.gotmpl", "*/*/*/*.gotmpl")
+	if err != nil {
+		return fmt.Errorf("instantiate template: %w", err)
 	}
 
-	protoDir := filepath.Join(fullPath, "proto")
-	if _, err := os.Stat(protoDir); errors.Is(err, os.ErrNotExist) {
-		fmt.Printf("Creating missing %q folder\n", protoDir)
-		if err := os.MkdirAll(protoDir, os.ModePerm); err != nil {
-			return fmt.Errorf("creating proto directory %v: %w", protoDir, err)
+	// create directories
+	for _, dir := range directories {
+		dirPath := filepath.Join(projectPath, dir)
+		if _, err := os.Stat(dirPath); errors.Is(err, os.ErrNotExist) {
+			fmt.Printf("Creating missing %q folder\n", dirPath)
+			if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
+				return fmt.Errorf("creating directory %v: %w", dirPath, err)
+			}
+		} else {
+			fmt.Println("src directory already exists, skipping")
 		}
-	} else {
-		fmt.Println("proto directory already exists, skipping")
 	}
 
-	abisDir := filepath.Join(fullPath, "abis")
-	if _, err := os.Stat(abisDir); errors.Is(err, os.ErrNotExist) {
-		fmt.Printf("Creating missing %q folder\n", abisDir)
-		if err := os.MkdirAll(abisDir, os.ModePerm); err != nil {
-			return fmt.Errorf("creating abis directory %v: %w", abisDir, err)
+	// create files
+	err = fs.WalkDir(templates, "templates", func(path string, d fs.DirEntry, err error) error {
+		if d.IsDir() ||
+			d.Name() == "externs.gotmpl" ||
+			d.Name() == "libGen.gotmpl" ||
+			d.Name() == "pb_mod.gotmpl" ||
+			d.Name() == "substreamsGen.gotmpl" ||
+			d.Name() == "mod.gotmpl" {
+			return nil
 		}
-	} else {
-		fmt.Println("abis directory already exists, skipping")
-	}
+		relativeEmbedPath := strings.TrimPrefix(path, "templates"+string(os.PathSeparator))
 
-	cargoTomlPath := filepath.Join(fullPath, "Cargo.toml")
-	if _, err := os.Stat(cargoTomlPath); errors.Is(err, os.ErrNotExist) {
-		if err := generate("Cargo.toml", tplCargoToml, g, cargoTomlPath); err != nil {
-			return fmt.Errorf("generating Cargo.toml: %w", err)
+		// Change duplicate template filenames
+		if d.Name() == "abimodfile.gotmpl" || d.Name() == "pb-modfile.gotmpl" {
+			relativeEmbedPath = relativeEmbedPath[:len(relativeEmbedPath)-17] + "mod.gotmpl"
```

Comment:
A `strings.Replace` seems more appropriate here, fiddling with the string at the character level makes it harder to "see" what is happening. Also, somehow changing the filename would maybe miss the fact that he needs to update this line.

While with a replace, it will be easier to spot (still place for mistake but less chance).